leo/tests/integration.rs

```diff
@@ -0,0 +1,377 @@
+            );
+        }
+        panic!()
+    }
+}
+
+fn copy_recursively(src: &Path, dst: &Path) -> io::Result<()> {
+    // Ensure destination directory exists
+    if !dst.exists() {
+        fs::create_dir_all(dst)?;
+    }
+
+    for entry in fs::read_dir(src)? {
+        let entry = entry?;
+        let file_type = entry.file_type()?;
+        let src_path = entry.path();
+        let dst_path = dst.join(entry.file_name());
+
+        if file_type.is_dir() {
+            copy_recursively(&src_path, &dst_path)?;
+        } else if file_type.is_file() {
+            fs::copy(&src_path, &dst_path)?;
+        } else {
+            panic!("Unexpected file type at {}", src_path.display())
+        }
+    }
+
+    Ok(())
+}
+
+/// Recursively compares the contents of two directories
+fn dirs_equal(dir1: &Path, dir2: &Path) -> io::Result<bool> {
+    let entries1 = collect_files(dir1)?;
+    let entries2 = collect_files(dir2)?;
+
+    // Check both directories have the same files
+    if entries1 != entries2 {
+        return Ok(false);
+    }
+
+    // Compare contents of each file
+    for relative_path in &entries1 {
+        let path1 = dir1.join(relative_path);
+        let path2 = dir2.join(relative_path);
+
+        let bytes1 = fs::read(&path1)?;
+        let bytes2 = fs::read(&path2)?;
+
+        if bytes1 != bytes2 {
+            return Ok(false);
+        }
+    }
+
+    Ok(true)
+}
+
+/// Collects all file paths relative to the base directory
+fn collect_files(base: &Path) -> io::Result<HashSet<PathBuf>> {
+    let mut files = HashSet::new();
+    for entry in walkdir::WalkDir::new(base).into_iter().filter_map(Result::ok) {
+        let path = entry.path();
+        if path.is_file() {
+            let rel_path = path.strip_prefix(base).unwrap().to_path_buf();
+            files.insert(rel_path);
+        }
+    }
+    Ok(files)
+}
+
+fn remove_all_in_dir(dir: &Path) -> io::Result<()> {
+    for entry in fs::read_dir(dir)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if path.is_dir() {
+            fs::remove_dir_all(path)?;
+        } else {
+            fs::remove_file(path)?;
+        }
+    }
+    Ok(())
+}
+
+fn snarkos_installed() -> bool {
+    Command::new("snarkos")
+        .arg("--version")
+        .stdout(Stdio::null())
+        .stderr(Stdio::null())
+        .status()
+        .is_ok_and(|status| status.success())
+}
+
+const SNARKOS_VERSION: &str = "4.0.1";
+
+fn install_snarkos() -> bool {
+    snarkos_installed() || {
+        println!("Installing snarkOS!");
+        Command::new("cargo")
+            .arg("install")
+            .arg("snarkos")
+            .arg("--version")
+            .arg(SNARKOS_VERSION)
+            .arg("--features")
+            .arg("test_network") // Enable the testing consensus heights.
+            .status()
+            .is_ok_and(|status| status.success())
+    }
+}
+
+fn run_snarkos_validator(i: usize, num_validators: usize) -> io::Result<Child> {
```

Comment:
Actually I would add `--dev-no-txs`. This will ensure validator 0 doesn't generate a bunch of test transactions, which eats up a lot of resources.